src/Random.mo

```diff
@@ -8,63 +18,206 @@ module {
+        bytes[2] := Nat8.fromNat(Nat64.toNat((n >> 16) & 0xFF));
+        bytes[3] := Nat8.fromNat(Nat64.toNat((n >> 24) & 0xFF));
+        bytes[4] := Nat8.fromNat(Nat64.toNat((n >> 32) & 0xFF));
+        bytes[5] := Nat8.fromNat(Nat64.toNat((n >> 40) & 0xFF));
+        bytes[6] := Nat8.fromNat(Nat64.toNat((n >> 48) & 0xFF));
+        bytes[7] := Nat8.fromNat(Nat64.toNat((n >> 56) & 0xFF));
+        Blob.fromArray(Array.fromVarArray(bytes))
+      }
+    )
   };
 
-  /// Uses entropy from the management canister with automatic resupply.
-  public func newAsync() : AsyncRandom {
-    todo()
+  /// Creates a random number generator suitable for cryptography
+  /// using entropy from the ICP management canister with automatic resupply.
+  public func crypto() : AsyncRandom {
+    AsyncRandom(func() : async* Blob { await rawRand() })
   };
 
   public class Random(generator : () -> Blob) {
+    var iter : Iter.Iter<Nat8> = Iter.empty();
+    let bitIter : Iter.Iter<Bool> = object {
+      var mask = 0x00 : Nat8;
+      var byte = 0x00 : Nat8;
+      public func next() : ?Bool {
+        if (0 : Nat8 == mask) {
+          switch (iter.next()) {
+            case null { null };
+            case (?w) {
+              byte := w;
+              mask := 0x40;
+              ?(0 : Nat8 != byte & (0x80 : Nat8))
+            }
+          }
+        } else {
+          let m = mask;
+          mask >>= (1 : Nat8);
+          ?(0 : Nat8 != byte & m)
+        }
+      }
+    };
 
     /// Random choice between `true` and `false`.
     public func bool() : Bool {
-      todo()
+      switch (bitIter.next()) {
+        case (?bit) { bit };
+        case null {
+          iter := generator().vals();
+          switch (bitIter.next()) {
+            case (?bit) { bit };
+            case null {
+              Runtime.trap("Random.bool(): generator produced empty Blob")
+            }
+          }
+        }
+      }
     };
 
     /// Random `Nat8` value in the range [0, 256).
-    public func byte() : Nat8 {
-      todo()
+    public func nat8() : Nat8 {
+      switch (iter.next()) {
+        case (?byte) { byte };
+        case null {
+          iter := generator().vals();
+          switch (iter.next()) {
+            case (?byte) { byte };
+            case null {
+              Runtime.trap("Random.nat8(): generator produced empty Blob")
+            }
+          }
+        }
+      }
     };
 
-    /// Random `Float` value in the range [0, 1).
-    /// The resolution is maximal (i.e. all relevant mantissa bits are randomized).
-    public func float() : Float {
-      todo()
+    // Helper function which returns a uniformly sampled `Nat64` in the range `[0, toExclusive)`.
+    // Uses rejection sampling to ensure uniform distribution even when the range
+    // doesn't divide evenly into 2^64. This avoids modulo bias that would occur
+    // from simply taking the modulo of a random 64-bit number.
+    func uniform64(toExclusive : Nat64) : Nat64 {
```

Comment:
```suggestion
    func uniform64(toExclusive : Nat64) : Nat64 {
```
Doesn't this mean you can't generate 2^64-1?
[0,0) is empty so the function should probably trap when toExclusive = 1.

Maybe this is case where we want an toInclusive, instead. Then no need to trap, and all values exressible.

Reply:
Sure; this makes sense given that it's a private function. It probably makes sense to keep `natRange()` exclusive while perhaps adding a `natRangeInclusive()` for consistency with the `*.range()` / `*.rangeInclusive()` naming convention.